keps/sig-node/3675-resource-plugin-manager/README.md

```diff
@@ -0,0 +1,1229 @@
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+The authors have taken inspiration from the [CSI development](https://github.com/kubernetes/design-proposals-archive/blob/main/storage/container-storage-interface.md). 
+Kubernetes compute management is tightly integrated with the Kubelet and the 
+existing suite of resources managers including the topology manager, CPU manager, 
+memory manager, and device manager. While these managers have added functionality that has addressed a varied set of use cases, they do present the community with several challenges.
+
+Adding a new capability to one of these managers is slow moving and difficult 
+due to their complex interactions, the level of prudence required given the potential
+impacts, the implied necessity to make the extension work in as many scenarios as
+possible, as well as the overhead heaped onto the sig-node maintainers. More 
+details on the challenges discussed with the community have been captured in 
+the [CPU Management Kubelet Use Cases & Current State document](https://docs.google.com/document/d/1U4jjRR7kw18Rllh-xpAaNTBcPsK5jl48ZAVo7KRqkJk).  
+
+As a result, adding optimizations necessary to support some more niche use cases, 
+to improve total cost of ownership for deployments with more exacting requirements 
+on resources or to provide a mechanism to progressively roll-out the benefits of 
+innovations related to newer CPU architectures is difficult and high cost to community resources.
+
+This proposal aims to address the challenges by introducing a new Container Compute 
+Interface (CCI). The CCI is conceptually equivalent to the CSI in that it will 
+delineate between the responsibility of the Kubelet and lower-level responsibility
+of compute focused plugins for resource assignment capabilities, such CPU and memory. 
+
+Given the existing Kubelet CPU/memory manager capabilities and their usefulness 
+for many scenarios, the proposal takes a minimally disruptive approach to
+introducing the CCI favouring an architecture that drops back to the existing
+behaviour seamlessly during error conditions.
```

Comment:
We take control completely for pods requiring CCI drivers, fall back will not be supported in current stage. Driver failures will lead to Pod failure. Still pods not requiring driver have to continue to work as before without impact from driver failure and cluster has to remain usable.

Reply:
My question wasn't about "who controls the pod". It's about "who controls the node resources".

Reply:
CPU and RAM decisions from a CCI Drivers are passed back to kubelet, , kubelet( CPU/Memory manager) applies those.

Reply:
I think that's the crux of the matter: in the traditional model, kubelet doesn't decide which pods get CPU and RAM on a node. That decision is made by kube-scheduler. The role of kubelet and the runtime is just to enforce that.

Reply:
Let me rephrase to avoid potential confusion: when I said "who controls the node resources", I meant "who controls resource assignment to pods during scheduling".

Reply:
CCI Drivers are executed after scheduling. Scheduling decisions have to be controlled by other components.

Reply:
That is a good non-goal then. Can you add it?

Reply:
added the scheduling part as  non-goal.

Reply:
resource ownership is explained in the policy section